### ISSUE TITLE
Correctness of summing up embeddings by the sequence dimension

### ISSUE THREAD

**Issue by koren-v** [cdpierse/transformers-interpret#56]
Hi, thanks for the great repo!

I would like to discuss the implementation of Integrated Gradients in combination with the text models. I found that you sum IG outputs by the last dimension to get a single scalar for each token, I also remember that the implementation of AllenNLP uses the [same](https://github.com/allenai/allennlp/blob/main/allennlp/interpret/saliency_interpreters/integrated_gradient.py#L33) trick. But do we really can simply sum them as the gradients can be negative? If some of the gradients for a particular token are negative, summing them with the rest will reduce the overall token's "score". Could it be more correct if we will take some kind of norm (L1, L2) for each token?

**Comment by cdpierse** [comment]
Hi @koren-v, thanks for your contribution here and for helping with some other issues around the repo, it's really appreciated and I'm really happy to see people exploring the package down to the implementation level.  

This is a very interesting discussion and I think I'm open to exploring parameterizing the type of embedding summarization that takes place **(sum, l1, l2)** as it can only add more certainty to model interpretations.

However I'm not fully convinced yet that summing a token's embedding gradients is less correct than taking the norm (l1 or l2) which seems to me like just another approach of representing an embedding as a single scalar. Could you possibly elaborate more on what you mean by correctness ?

My reasoning against this is that if a token's embedding has negative values mixed in with positive values this should not just be assumed to be noise that is reducing the token's true score, the negative values could well be indicative that there is less certainty of a token's positive contribution for a given sequence, they are a feature not a bug. Assuming that a transformer model is well trained etc I like to think that produced embeddings are rich and adequately represent a token's contribution to a sequence.

I put together a small but by no means comprehensive example below. I'm using the exact same code snippet that I use for the example of a classification explainer, for the sequence `I love you, I like you`. I wanted to explore very simply the relation between the words `love` and `like` when using l2 norm (pytorch default I believe) vs simply summing. The results can be seen below and you should be able to replicate the same. 

My main observation is that when summing embedding values the word `love` is ~4x greater than `like` while when we l2 norm the same tokens `love` is ~ 1.7x greater than `like`. My gut intuition here is to favor the representation that amplifies the magnitude of the word `love` over `like` as the this is a sentiment classification model. Of course this is just my observation for this example but I think it might hold true in general. 


```python

cls_explainer("I love you, I like you")

In [53]: cls_explainer.attributions._attributions[0][2].norm() #l2 norm for love
Out[53]: tensor(0.0574, dtype=torch.float64)

In [54]: cls_explainer.attributions._attributions[0][2].sum() # sum for love
Out[54]: tensor(0.3952, dtype=torch.float64)

In [55]: cls_explainer.attributions._attributions[0][6].norm()  #l2 norm for like
Out[56]: tensor(0.0332, dtype=torch.float64)

In [56]: cls_explainer.attributions._attributions[0][6].sum() # sum for like
Out[56]: tensor(0.0968, dtype=torch.float64)

```

Like I said above though I think adding normalization as an optional alternative for producing scalar values for token attribution values is worth exploring and possibly adding to the package. Would you be possibly interested in working with me on a PR for this, it would mostly be adding a new summarize method to the LIGAttributions class and some minor work on explainer's API's.

**Comment by koren-v** [comment]
@cdpierse "if a token's embedding has negative values mixed in with positive values this should not just be assumed to be noise that is reducing the token's true score, the negative values could well be indicative that there is less certainty of a token's positive contribution for a given sequence" - I'm not certainly sure that the sign of the gradient can be important in this case, I believe that just the value itself means something. Because of that, I assumed that taking L1/L2 norm should be more "correct".

**Comment by cdpierse** [comment]
Hi @koren-v , I have a feeling that in this case of the embedding attributions the sign itself is important as the sum of a attribution embedding decides whether or not it contributes positively or negatively towards a prediction, so performing L1/L2 would potentially hide or negate this importance.

Such regularization has a much clearer objective when training a model but in attribution and explainability it would stop us from being able to understand values that _negatively_ contribute and would in fact incorrectly invert those values as being _positive_ contributors. For example, in the sentiment classification below the value of the word `dislike` when positively signed would incorrectly suggest it contributes to the output.
<img width="706" alt="Screenshot 2021-07-05 at 16 27 10" src="https://user-images.githubusercontent.com/8831892/124493718-de160d80-ddad-11eb-97aa-d3b9611eedf4.png">



That being said I'm not an expert on the summarization of these attributions and I took my inspiration from this [example ](https://github.com/pytorch/captum/blob/master/tutorials/Bert_SQUAD_Interpret.ipynb)from the team at Captum. 

I did have another quick read over the [Axiomatic Attribution for Deep Networks ](https://arxiv.org/abs/1703.01365) paper and it would seem that when explaining token level embeddings the sign in the attribution embedding is important. In that a negatively or positively signed scalar in an attribution embedding tries to approximate which dimensions of the embedding itself are contributing to the given output as if the embedding's scalars were feature inputs.

I hope this has cleared it up somewhat, if not I'd recommend maybe raising a similar issue on the [Captum Github](https://github.com/pytorch/captum) they have been very helpful and fast to respond to questions that I have had in the past.

**Comment by koren-v** [comment]
Thanks for your answer! 
Finally, I reread the paper and got that had a wrong assumption about the usage of the L1/L2 norm.
In addition, I found great [howto](https://github.com/ankurtaly/Integrated-Gradients/blob/master/howto.md) by the authors of the paper, suggest checking it and maybe, refer to it in your readme)
